in-progress/8754-slashing-staking/design.md

```diff
@@ -0,0 +1,436 @@
+
+> [!info] Top-ups  
+> We are not supporting "topping up" the account, since it introduce a few additional code paths that we simple are not interested in for this minimal setup.
+
+## Users and Interaction
+
+The users of the staking contract are:
+
+- The rollup instance, to sample from the set
+- The slasher, to slash operators
+- Actors that wish to add another operator to the set, exit or vote on slashing
+
+The implementation itself can be made as part of the rollup instance, or as a separate contract, for now we are just going with the same contract as it makes some of the data access slightly simpler as we don't need to provide as many getters.
+
+For users indending to be validators, there is mainly three points of interest:
+
+- Joining the set
+- Exiting the set
+- Voting on slashing
+
+Joining and exiting is done directly on the contract, likely using a frontend or CLI but connected to a cold-storage wallet. These actions are seldom and deal with value so it is important to keep it safe.
+
+As mentioned, the voting on slashing is to be done by the node itself by "slashing-modules". Essentially a small pieces of code that the node is running that will identify misbehaviour and submit or vote on slashing proposal. Votes are collected in the same manner as the `GovernanceProposer` used for governance proposals, with the caveat that if it is "to be proposed" it will go to the slasher where it becomes executable - it does not need to pass a normal governance vote.
+
+### Delegation
+
+As the `withdrawer` can be a completely separate actor from the operator, it can be used directly as a "poor mans" delegation, or using extra contracts for a more sophisticated delegation.
+An example of the latter would be a LST contract where the `withdrawer` is the LST contract itself but the operator is some node operator like LIDO.
+
+## Implementation
+
+The staking contract will be utilising a few maps and sets to keep track of validators and their states. The primary interest is the `attesters` set which we can use for dealing with the currently "active" validators, e.g., people that might be chosen for committees. We use the enumerable set to have a neat birectional map, making it simple for us to go from `index => attester` and vice versa.
+
+We have a few states where it is allowed to have an operator listed, even though it is not active, so we keep track of the status as part of the `OperatorInfo`.
+
+```solidity
+import {ERC20, SafeERC20} from "...";
+import {Timestamp} from "...";
+import {EnumerableSet}
+
+// None -> Does not exist in our setup
+// Validating -> Participating as validator
+// Living -> Not participating as validator, but have funds in setup,
+// 			 hit if slashes and going below the minimum
+// Exiting -> In the process of exiting the system
+enum ApeStatus { NONE, VALIDATING, LIVING, EXITING }
+
+struct OperatorInfo {
+	uint256 stake;
+	address withdrawer;
+	address proposer;
```

Comment:
bytes32 for proposer?

Reply:
![image](https://github.com/user-attachments/assets/6f4b026c-bac9-4535-b46f-7529d2b05314)